src/main/java/org/springframework/authz/server/config/SecurityConfigurer.java

```diff
@@ -0,0 +1,48 @@
+package org.springframework.authz.server.config;
+
+import org.springframework.authz.server.filter.OAuthGrantBasedAuthenticationFilter;
+import org.springframework.authz.server.filter.matcher.GrantTypeReqMatcher;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.config.BeanIds;
+import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.authentication.www.BasicAuthenticationConverter;
+import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+
+@EnableWebSecurity
+public class SecurityConfigurer extends WebSecurityConfigurerAdapter {
+	
+	@Autowired
+	private OAuthGrantBasedAuthenticationFilter clientCredentialsFilter;
+	
+	@Autowired
+	private RequestMatcher clientCredReqGrantMatcher;
+	
+	@Override
+    protected void configure(HttpSecurity http) throws Exception {
+    	
+    	http.authorizeRequests()
+				.antMatchers("/token").permitAll();
+    	
+    	http.addFilterBefore(clientCredentialsFilter, BasicAuthenticationFilter.class);

```

Comment:
You are not validating the client credentials yourself which leaves me to the conclusion that you are relying on the BasicAuthenticationFilter from the spring-security-filterchain.
This is actually a problem. RFC6749 defines an additional operation to the basic authenticaton scheme that is not directly mentioned in RFC7617 (Basic Auth Scheme). This causes problems with libraries like nimbus that do honor this additional operation.
https://tools.ietf.org/html/rfc6749#section-2.3.1
Simplified, if a client uses the basic authentication scheme the client_id parameter and the client_secret must be url-encoded. the BasicAuthenticationFilter of spring does not decode such authentication details and thus the authentication will fail.

Reply:
Actually I am not relying on the BasicAuthorizationFilter in my flow. The configuration of the client credentials filter that I have developed could be seen in the AuthorizationFilterConfigurations.java  in my PR. But for extracting the client id and secret from authorization header, I am using BasicAuthenticationConverter which I can replace with a custom class and inject.

As far as the basic auth is concerned, current implementation will extract the authorization header, make sure its Basic, then decode the base64 encoded string and split the result by colon to obtain client id and secret.

Now for the missing piece, do you mean that client id and secret obtained from header should further be url-decoded or do you mean that client id and secret passed in request body should be url-decoded ? If you mean the latter then that is not yet developed. I am a bit confused because the spec link you mentiond specifies application/x-www-form-urlencoded which is more relevant to content of the body rather the header (but correct me if I am wrong). 

Once that is clear I can make changes accordingly.

Thanks for your inputs.

Paurav.

Reply:
I mean the base64 decoded client_id and client_secret from the header. I recently made the experience that if this is not done it might come to a clash with the nimbus library. As pointed out in section 2.3.1 of RFC6749 the client_id and client_secret must be encoded.

> Clients in possession of a client password MAY use the HTTP Basic
>    authentication scheme as defined in [RFC2617] to authenticate with
>    the authorization server.  The client identifier is encoded using the
>    "application/x-www-form-urlencoded" encoding algorithm per
>    Appendix B, and the encoded value is used as the username; the client
>    password is encoded using the same algorithm and used as the
>    password.

and then from https://www.w3.org/TR/1999/REC-html401-19991224/interact/forms.html#h-17.13.4.1

>  application/x-www-form-urlencoded  
>
> This is the default content type. Forms submitted with this content type must be encoded as follows:
> 
>  1. Control names and values are escaped. Space characters are replaced by `+', and then reserved characters are escaped as described in [RFC1738], section 2.2: Non-alphanumeric characters are replaced by `%HH', a percent sign and two hexadecimal digits representing the ASCII code of the character. Line breaks are represented as "CR LF" pairs (i.e., `%0D%0A').
> 2. The control names/values are listed in the order they appear in the document. The name is separated from the value by `=' and name/value pairs are separated from each other by `&'.

at least the nimbus developers implemented it this way and I had a short discussion with them about this topic and agreed eventually on their solution